Run various queries from other queries instead of explicitly in phases

oli-obk opened the pull request:
These are just legacy leftovers from when rustc didn't have a query system. While there are more cleanups of this sort that can be done here, I want to land them in smaller steps.

This phased order of query invocations was already a lie, as any query that looks at types (e.g. the wf checks run before) can invoke e.g. const eval which invokes borrowck, which invokes typeck, ...

oli-obk committed:
Allow `LocalDefId` as the argument to `def_path_str`

oli-obk committed:
Make `check_match` and `check_liveness` take a `LocalDefId`

oli-obk committed:
Run `check_match` and `check_liveness` when MIR is built instead of having an explicit phase for them

oli-obk commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit a1bf293aea5384252c0d7227f46dcf534667aeb1 with merge 41c8d3a20cad3d4de552339500551ada93f0d722...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a1bf293aea5384252c0d7227f46dcf534667aeb1","merge_sha":"41c8d3a20cad3d4de552339500551ada93f0d722"} -->

Zoxc commented on compiler/rustc_mir_build/src/build/mod.rs:
```diff
@@ -71,6 +72,16 @@ fn mir_build(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_
             let thir = thir.steal();
 
+            if def.const_param_did.is_none() {
+                parallel!({ tcx.ensure().check_match(def.did) }, {
```
I would avoid parallelism deeper in the query system to avoid the deadlock risk that brings.

bjorn3 commented on compiler/rustc_codegen_cranelift/scripts/filter_profile.rs:
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env bash
-#![forbid(unsafe_code)]/* This line is ignored by bash
+#![forbid(unsafe_code)]
+/* This line is ignored by bash
```
Please revert this change. It is necessary for being both valid rust code and valid bash script.

oli-obk replied:
oh huh... rustfmt changed this.

bjorn3 replied:
It is in the ignore section of `rustfmt.toml`, but rust-analyzer doesn't respect this.

bjorn3 commented on compiler/rustc_hir_analysis/src/lib.rs:
```diff
@@ -529,8 +529,6 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
         tcx.hir().for_each_module(|module| tcx.ensure().check_mod_item_types(module))
     });
 
-    tcx.sess.time("item_bodies_checking", || tcx.typeck_item_bodies(()));
```
What is going to ensure that typecheck errors for all functions are reported without this call?

bjorn3 replied:
This also regresses the quality of the output of `-Ztime-passes`.

oli-obk replied:
as mentioned in the main post

> This phased order of query invocations was already a lie, as any query that looks at types (e.g. the wf checks run before) can cause e.g. const eval which invokes borrowck, which invokes typeck, ...

we already didn't produce correct time tracking for it

jackh726 commented:
I think this is a good change, but maybe a safer intermediate step would be to "assert" that all the queries were run in the "passes" for now. This would ensure that we don't accidentally make a query more lazy, then skip something (accidentally).

oli-obk commented:
Hmm... we don't have the setup for doing that (also doing it at that phase won't work, as it may get run later). I'll see if I can add something to the query system for asserting that a query has been run

bors commented:
:umbrella: The latest upstream changes (presumably #101841) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
> This would ensure that we don't accidentally make a query more lazy, then skip something (accidentally).

I looked into it, and we're already relying on the `mir_built` query to run `thir_check_unsafety`. So we're already assuming that `mir_built` gets run for all things with bodies. I agree some assertions would be neat, but generating a check-function per query and calling it is a lot. Limiting it to certain queries can be done with more query flags, but that seems like a lot of infrastructure for something that we can be certain won't happen (as we'd already be missing checks in that case).

oli-obk committed:
Leave it to the query system to invoke the typeck query instead of invoking it eagerly.

Later queries that are run on all body owners will invoke typeck as they need information from its result to perform their own logic

oli-obk commented:
@rust-timer build 41c8d3a20cad3d4de552339500551ada93f0d722

rust-timer commented:
Queued 41c8d3a20cad3d4de552339500551ada93f0d722 with parent 639377ed737b25830ec44dc6acf93467c980316a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=639377ed737b25830ec44dc6acf93467c980316a&end=41c8d3a20cad3d4de552339500551ada93f0d722).
<!-- rust-timer: temporary -->

cjgillot commented:
I'm not sure I agree with the general direction. I'd rather suggest doing the opposite:
- remove useless dependencies from queries -> avoid useless recomputations;
- invoke all the required "check" queries from `analysis` -> ensure correctness-related stuff are in a single place.

The fact that we run `thir_check_unsafety` from `mir_built` is more accidental, due to the fact that `mir_built` steals THIR. We probably shouldn't really rely on it for correctness. Even if having `ensure().mir_borrowck` in `analysis` is enough to make sure that `mir_built` is computed and `thir_check_unsafety` too.

oli-obk commented:
The problem with those useless dependencies is that e.g. const eval can run on code that doesn't satisfy various checks, leading to surprising bugs and workarounds in const eval. I agree that mir_built may be too early, but at the very least the mir_drops_elaborated_and_const_checked query should not finish before the various checks on the same def id are done.

I guess my assumption was that at some point all we have to do is to either call the optimized_mir query for the main function and can compile from just the information from that, or call the analysis query to do a check build.

If changing any of this negatively affects incremental, then we should not do it, but I don't see that being an issue in general with this scheme, just with the details of where to add edges.

rust-timer commented:
Finished benchmarking commit (41c8d3a20cad3d4de552339500551ada93f0d722): [comparison URL](https://perf.rust-lang.org/compare.html?start=639377ed737b25830ec44dc6acf93467c980316a&end=41c8d3a20cad3d4de552339500551ada93f0d722&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=639377ed737b25830ec44dc6acf93467c980316a&end=41c8d3a20cad3d4de552339500551ada93f0d722&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.7%, 0.7%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.7%, 0.7%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.6%, -0.5%] | 6     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 1     |
| All ❌✅ (primary)                 | -0.4% | [-0.6%, 0.7%]  | 7     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=639377ed737b25830ec44dc6acf93467c980316a&end=41c8d3a20cad3d4de552339500551ada93f0d722&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1%  | [1.1%, 1.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [1.4%, 2.8%]   | 10    |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.5%, -1.0%] | 3     |
| Improvements ✅ <br /> (secondary) | -3.9% | [-7.1%, -2.5%] | 5     |
| All ❌✅ (primary)                 | -0.9% | [-2.5%, 1.1%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=639377ed737b25830ec44dc6acf93467c980316a&end=41c8d3a20cad3d4de552339500551ada93f0d722&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-1.7%, -1.6%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.5%, -2.0%] | 2     |
| All ❌✅ (primary)                 | -1.6% | [-1.7%, -1.6%] | 2     |
</details>

oli-obk commented:
Everything but `assoc_many_items` in `doc` mode is noise.

oli-obk commented:
> * remove useless dependencies from queries -> avoid useless recomputations;

about that, do we really add more dependencies with `ensure`? Doesn't `ensure` just mean that the ensured query must be run before the current one without any change in the dep-graph otherwise?

cjgillot commented:
Yes, `ensure` adds a dependency edge. And the calls to `ensure` just before stealing a value bother me a bit: they introduce a dependency to a value that will not be used, and don't even make sure the code will not be called later. `ensure` may "just" mark the query as green, without loading any value, to be recomputed later when the value is actually required.

> The problem with those useless dependencies is that e.g. const eval can run on code that doesn't satisfy various checks, leading to surprising bugs and workarounds in const eval. I agree that mir_built may be too early, but at the very least the mir_drops_elaborated_and_const_checked query should not finish before the various checks on the same def id are done.

In those cases, we should rely on the `trainted_by_errors` fields on typeck and MIR body.

One other symptom is the calls to `track_errors` in `rustc_hir_analysis::check_crate` that shouldn't exist: typeck should rely on propagated error state via `ty::Error` and `ty::ReError`. But that FIXME exists since 2020, so that's not high priority either.

> If changing any of this negatively affects incremental, then we should not do it, but I don't see that being an issue in general with this scheme, just with the details of where to add edges.

From the perf results, I have no proof that it does negatively affect incremental, so no serious grounds on which to block this PR. That's also why I frame it as a design question, more than an operational question.

That said, we should merge that PR, as it does not lock anything in a way or the other. I'll have to make my case with an actual PR demonstrating the direction I have in mind :smile:.

oli-obk commented:
> In those cases, we should rely on the `trainted_by_errors` fields on typeck and MIR body.

We do, but unsafety checks run in a query, and now, that query is actually invoked before const eval runs. Before this PR that was not guaranteed.

oli-obk commented:
> That said, we should merge that PR, as it does not lock anything in a way or the other. I'll have to make my case with an actual PR demonstrating the direction I have in mind 😄.

Well, this PR isn't in a rush, and I don't want to make things harder for you. If we land this one we'll land more afterwards. Could this cause a soft lock-in that will make it hard/impossible for you to later change it to reduce the number of dependency edges?

bors commented:
:umbrella: The latest upstream changes (presumably #108301) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit fd6586f399bafbf5c72346cf89d119fe877376ef with merge dcb744f3ea5fd667ba604d670360ecb581f054ea...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fd6586f399bafbf5c72346cf89d119fe877376ef","merge_sha":"dcb744f3ea5fd667ba604d670360ecb581f054ea"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4246510232/jobs/7385523248)
Build commit: dcb744f3ea5fd667ba604d670360ecb581f054ea (`dcb744f3ea5fd667ba604d670360ecb581f054ea`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4246510232/jobs/7385523248"},"merge_sha":"dcb744f3ea5fd667ba604d670360ecb581f054ea"} -->

rust-timer commented:
Queued dcb744f3ea5fd667ba604d670360ecb581f054ea with parent b869e84e581612f4a30a4bca63bd9e90e9a17003, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b869e84e581612f4a30a4bca63bd9e90e9a17003&end=dcb744f3ea5fd667ba604d670360ecb581f054ea).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (dcb744f3ea5fd667ba604d670360ecb581f054ea): [comparison URL](https://perf.rust-lang.org/compare.html?start=b869e84e581612f4a30a4bca63bd9e90e9a17003&end=dcb744f3ea5fd667ba604d670360ecb581f054ea&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=b869e84e581612f4a30a4bca63bd9e90e9a17003&end=dcb744f3ea5fd667ba604d670360ecb581f054ea&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.2%, 2.6%]   | 5     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.6%] | 6     |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.6%, -0.2%] | 2     |
| All ❌✅ (primary)                 | -0.6% | [-0.6%, -0.6%] | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b869e84e581612f4a30a4bca63bd9e90e9a17003&end=dcb744f3ea5fd667ba604d670360ecb581f054ea&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.4%  | [4.0%, 4.9%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-4.6%, -2.6%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b869e84e581612f4a30a4bca63bd9e90e9a17003&end=dcb744f3ea5fd667ba604d670360ecb581f054ea&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.8%, 1.1%]   | 5     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [2.4%, 3.1%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.8% | [-6.6%, -4.9%] | 6     |
| All ❌✅ (primary)                 | 0.9%  | [0.8%, 1.1%]   | 5     |
</details>

cjgillot commented:
Seeing https://github.com/rust-lang/rust/pull/108705, I finally convinced myself that this PR is the right direction. This can help to avoid building MIR for broken code, which is that much potential ICEs and non-sensical diagnostics.

> https://github.com/rust-lang/rust/pull/108118#issuecomment-1433083909
> Hmm... we don't have the setup for doing that (also doing it at that phase won't work, as it may get run later). I'll see if I can add something to the query system for asserting that a query has been run

That's what `ensure()` does: either the query is known green and they replay diagnostics from the previous compilation, or the query is re-executed.

oli-obk commented:
> That's what `ensure()` does: either the query is known green and they replay diagnostics from the previous compilation, or the query is re-executed.

Oh I meant something that asserts it instead of executing it. Though ensuring later will ICE due to stolen stuff, so it results in the same thing

bors commented:
:umbrella: The latest upstream changes (presumably #108089) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4755725809/jobs/8450274572) [(plain)](https://github.com/rust-lang/rust/commit/587b5eac715dd6337e88c7c56c749b4f4d0601ac/checks/12899135858/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......................ii........i..................................

failures:

---- [mir-opt] tests/mir-opt/const_promotion_extern_static.rs stdout ----
18 -         StorageLive(_3);                 // scope 0 at $DIR/const_promotion_extern_static.rs:+0:31: +0:46
19 -         StorageLive(_4);                 // scope 0 at $DIR/const_promotion_extern_static.rs:+0:32: +0:45
20 -         StorageLive(_5);                 // scope 1 at $DIR/const_promotion_extern_static.rs:+0:42: +0:43
- -         _5 = const {alloc2: *const i32}; // scope 1 at $DIR/const_promotion_extern_static.rs:+0:42: +0:43
+ -         _5 = const {alloc3: *const i32}; // scope 1 at $DIR/const_promotion_extern_static.rs:+0:42: +0:43
22 +         _6 = const _;                    // scope 0 at $DIR/const_promotion_extern_static.rs:+0:31: +0:55
23                                            // mir::Constant
24 -                                          // + span: $DIR/const_promotion_extern_static.rs:13:42: 13:43

- -                                          // + literal: Const { ty: *const i32, val: Value(Scalar(alloc2)) }
+ -                                          // + literal: Const { ty: *const i32, val: Value(Scalar(alloc3)) }
26 -         _4 = &(*_5);                     // scope 1 at $DIR/const_promotion_extern_static.rs:+0:41: +0:43
27 -         _3 = [move _4];                  // scope 0 at $DIR/const_promotion_extern_static.rs:+0:31: +0:46
28 -         _2 = &_3;                        // scope 0 at $DIR/const_promotion_extern_static.rs:+0:31: +0:55
50       }
51   }
52 - 
52 - 
- - alloc2 (extern static: X)
+ - alloc3 (extern static: X)
55 


thread '[mir-opt] tests/mir-opt/const_promotion_extern_static.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/const_promotion_extern_static.FOO.PromoteTemps.diff', src/tools/compiletest/src/runtest.rs:3527:21


failures:
    [mir-opt] tests/mir-opt/const_promotion_extern_static.rs
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #96840) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 334423263a347b38ed0040be788feb3a4197c056 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 334423263a347b38ed0040be788feb3a4197c056 -->
<!-- homu: {"type":"Approved","sha":"334423263a347b38ed0040be788feb3a4197c056","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 334423263a347b38ed0040be788feb3a4197c056 with merge 3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f...
<!-- homu: {"type":"BuildStarted","head_sha":"334423263a347b38ed0040be788feb3a4197c056","merge_sha":"3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4778626795/jobs/8496023100)
Approved by: cjgillot
Pushing 3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4778626795/jobs/8496023100"},"merge_sha":"3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4778626795/jobs/8496023100)
Approved by: cjgillot
Pushing 3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4778626795/jobs/8496023100"},"merge_sha":"3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f"} -->

rust-timer commented:
Finished benchmarking commit (3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f): [comparison URL](https://perf.rust-lang.org/compare.html?start=9e540df7931a32ca286eb2e907afde78b718fd17&end=3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9e540df7931a32ca286eb2e907afde78b718fd17&end=3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.5%, 0.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.2%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9e540df7931a32ca286eb2e907afde78b718fd17&end=3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.7% | [-4.4%, -1.6%] | 6     |
| Improvements ✅ <br /> (secondary) | -1.4% | [-1.4%, -1.4%] | 1     |
| All ❌✅ (primary)                 | -2.7% | [-4.4%, -1.6%] | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9e540df7931a32ca286eb2e907afde78b718fd17&end=3462f79e94f466a56ddaccfcdd3a3d44dd1dda9f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6% | [1.6%, 1.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | 3.8% | [3.8%, 3.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.6% | [1.6%, 1.6%] | 1     |
</details>
